076-moving-data-between-two-jbod-disks-using-cruise-control.md

```diff
@@ -0,0 +1,89 @@
+
+## Motivation
+
+We should introduce the logic to Strimzi to leverage Cruise Control integration and make it possible to move the data between two JBOD disks.
+This feature will also allow us to remove the disks without the loss of data.
+
+## Proposal
+
+Cruise Control provides `remove_disks` HTTP REST endpoint to move replicas from a specified disk to other disks of the same broker. It always uses intra-broker re-balancing.
+This endpoint triggers a rebalancing operation by moving replicas in a size-based manner to the remaining disks, from the largest to the smallest, while checking the following constraint:
+```sh
+1 - (remainingUsageAfterRemoval / remainingCapacity) > errorMargin
+```
+where:
+```sh
+remainingUsageAfterRemoval = current usage for remaining disks + additional usage from removed disks
+remainingCapacity = sum of capacities of the remaining disks
+errorMargin = configurable property (default 0.1); it makes sure that a disk percentage is always free when moving replicas
+```
+
+In order to use the `remove_disks` endpoint in the Strimzi cluster operator, it would be added to the [`CruiseControlApi`](https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlApi.java) interface and developing the corresponding implementation.
+
+### Implementation
+
+To implement this feature, we will be adding a new mode to the `KafkaRebalanceMode` class.
+* `remove-disks`: It moves replicas from a specified disk to other disks of the same broker. It always uses intra-broker re-balancing.
+You can use this mode by changing the `spec.mode` to `remove-disks` in the `KafkaRebalance` resource.
+
+A `KafkaRebalance` custom resource would look like this.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  # setting the mode as `remove-disks` to move data between the JBOD disks
+  mode: remove-disks
+  # providing the list of brokers, and the corresponding volumes from which you want to move the replicas
+  moveReplicasOffVolumes:
+    - brokerId: 0
+      volumeIds: [1, 2]
+    - brokerId: 2
+      volumeIds: [1]
+# ...
+```
+
+### Flow
+
+- The user should be using the `Kafka` resource with JBOD configured. Make sure that you have more than one disk configured on the brokers.
+- When the Kafka cluster is ready, the user creates a `KafkaRebalance` custom resource with the `spec.mode` field as `remove-disks` and provides list of the brokers, and the corresponding volumes from which you want to move the replicas in the `spec.moveReplicasOffVolumes` field. In case, the `spec.moveReplicasOffVolumes` field is not set, then the `KafkaRebalance` resource will move to `NotReady` state prompting that `spec.moveReplicasOffVolumes` field is missing.
```

Comment:
```suggestion
- When the Kafka cluster is ready, the user creates a `KafkaRebalance` custom resource with the `spec.mode` field as `remove-disks` and provides a list of the brokers, and the corresponding volumes from which they want to move the replicas in the `spec.moveReplicasOffVolumes` field. In case, the `spec.moveReplicasOffVolumes` field is not set, then the `KafkaRebalance` resource will move to `NotReady` state prompting that `spec.moveReplicasOffVolumes` field is missing.
```